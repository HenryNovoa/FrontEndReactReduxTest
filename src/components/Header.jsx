import React from 'react'

const Header = props => {
    return (
        <div className="section">
            <h1 className="title">STARWARS</h1>
            <p className="subtitle">
                Search for your favorite star wars vehicles and starships and see their length in meters!
      </p>
            {props.children}
        </div>
    )
}

export default Header