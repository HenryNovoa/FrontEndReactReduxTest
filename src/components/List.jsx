import React from 'react'
import { connect } from 'react-redux'
const mapStateToProps = state => {
    return { remoteItems: state.remoteItems}
}

export const ConnectedList = (props) => (
    <div className='section'>
        <h1 className='title'>Search Results</h1>
        <ul className='content'>
            {props && props.remoteItems && props.remoteItems.length ?
               props && props.remoteItems && props.remoteItems.map((el, index) => (
                <li className='box' key={index}>
                    <div className='columns'>
                        <div className='column'>
                            {`name:${el.name} `}
                        </div>
                        <div className='column'>
                            {`model: ${el.model}`}
                        </div>
                        <div className='column'> 
                            {`length: ${el.length} meters`}
                        </div>
                    </div>
                </li>
            )):
            <li className='box'>
                    <div className='columns'>
                        <div className='column'>
                            {'No Search Results'}
                        </div>
                    </div>
                </li>
            }
        </ul>
    </div>
)
const List = connect(mapStateToProps)(ConnectedList)
export default List

