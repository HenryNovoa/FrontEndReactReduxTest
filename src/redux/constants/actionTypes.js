// String mistypes can lead to hard to find bugs, they are stored in variables to solved this

export const DATA_FETCH = 'DATA_FECTH'