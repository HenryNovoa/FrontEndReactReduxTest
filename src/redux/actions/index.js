import { call } from '../../fetch/index'
import { DATA_FETCH } from '../constants/actionTypes'

/**
 * 
 * @param {String} parameters The items to search for in the API
 * 
 * @returns {Promise} Calls the global function and fetches for data
 */
export const getData = parameters => {
    const { path, query } = parameters
    return dispatch => {
        return call({
            path: `${path}/?search=${query}`,
            method: 'GET',
            dispatch,
            actionType: DATA_FETCH
        })
    }
}