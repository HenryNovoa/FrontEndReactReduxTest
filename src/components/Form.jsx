import React, { useState }  from 'react'
import { connect } from "react-redux"
import { getData } from '../redux/actions/index'

function mapDispatchToProps(dispatch) {
    return {
        GET_DATA: item => dispatch(getData(item))
    }
}

const mapStateToProps = state => {
    return { vehicleTypes: state.vehicleTypes }
}

export const ConnectedForm = props => {

    const [searchInput, setSearchInput] = useState()
    const [vehicleTypeSearch, setVehicleTypeSearch] = useState()
    //state that manages if an API call is being made and sets the button for loading
    const [isLoading, setIsLoading] = useState(false)

    const handleInputChange = event => {
        const {
            target: {
                value
            }
        } = event

        setSearchInput(value)
    }

    const handleSelectChange = event => {
        const {
            target: {
                value
            }
        } = event
        setVehicleTypeSearch(value)
    }

    const handleSubmit = event => {
        event.preventDefault()
        setIsLoading(true)
        if (areInputsValid()) {
            props.GET_DATA({
                query: searchInput,
                path: vehicleTypeSearch
            })
            .then(()=>{
                setIsLoading(false)
            })
        }
    }

    const areInputsValid = () => {
        if (searchInput && vehicleTypeSearch) {
            return true
        }
        return false
    }

    const handleClearSearch = event => {
        event.preventDefault()
        setSearchInput('')
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className='field'>
                <div className='control'>
                    <input onChange={handleInputChange} 
                        value={searchInput || ''} 
                        className='input' 
                        type='text' 
                        placeholder='Vehicle or Spaceship' />
                </div>
            </div>

            <div className='field'>
                <p className='control'>
                    <span className='select'>
                        <select onChange={handleSelectChange}>
                            <option value=''>Select</option>
                            {props && props.vehicleTypes && props.vehicleTypes.map((type, index) => {
                                return (
                                    <option key={index} value={type}>{type}</option>
                                )
                            })}
                        </select>
                    </span>
                </p>
            </div>

            <div className='buttons'>
                <button type='submit'
                disabled={!areInputsValid()}
                className={isLoading ? `button is-primary is-loading` : 'button is-primary'}>Search</button>
                <button onClick={handleClearSearch} className='button is-link'>Clear</button>
            </div>
        </form>
    )
}

const Form = connect(mapStateToProps, mapDispatchToProps)(ConnectedForm)
export default Form
