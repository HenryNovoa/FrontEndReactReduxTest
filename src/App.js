import React from 'react'
import './App.sass'
import Header from './components/Header'
import Form from './components/Form'
import List from './components/List'

function App() {
  return (
    <>
    <Header>
      <Form/>
    </Header>
    <List/>
    </>
  )
}

export default App
