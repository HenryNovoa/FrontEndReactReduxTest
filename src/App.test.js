import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import { Provider } from 'react-redux'
import store from './redux/store/index'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import fetchMock from 'fetch-mock'
import * as types from './redux/constants/actionTypes' //'../../constants/ActionTypes'
import * as actions from './redux/actions/index'//'../../actions/TodoActions'
import rootReducer, { initialState } from './redux/reducers/index'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Header from './components/Header'
import toJson from 'enzyme-to-json';
import { ConnectedForm } from './components/Form'
import { ConnectedList } from './components/List'
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)


describe('default test', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Provider store={store}>
      <App />
    </Provider>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
})


describe('state management - redux', () => {
  describe('async actions', () => {
    afterEach(() => {
      fetchMock.restore()
    })
    it('should create a FETCH_DATA action when fetching data has been done', () => {
      fetchMock.getOnce('https://swapi.co/api/starships/?search=death', {
        body: {
          results: [{
            name: 'Deathstar',
            model: 'DS-1 Orbital Battle Station',
            length: '120000'
          }
          ]
        }
      })
        .catch(err => {
          throw new Error(err)
        })
      const expectedActions = [
        {
          type: types.DATA_FETCH,
          payload: [{
            name: 'Deathstar',
            model: 'DS-1 Orbital Battle Station',
            length: '120000'
          }
          ]
        }
      ]
      const store = mockStore({ remoteItems: [] })

      return store.dispatch(actions.getData({
        path: 'starships',
        query: 'death'
      })).then(() => {
        // return of async actions
        expect(store.getActions()).toEqual(expectedActions)
      })
    })
  })
  describe('reducers', () => {
    it('should return the initial state', () => {
      expect(rootReducer(undefined, {})).toEqual(initialState)
    })

    it('should replace remoteItems state with given payload on handle DATA_FETCH', () => {
      expect(
        rootReducer(undefined, {
          type: types.DATA_FETCH,
          payload: {
            test: 'testing'
          }
        })
      ).toEqual(
        {
          ...initialState,
          remoteItems: {
            test: 'testing'
          }
        }
      )
    })
  })
})

describe('React components', () => {
  Enzyme.configure({ adapter: new Adapter() })

  describe('Header', () => {

    function setup() {
      const enzymeWrapper = shallow(<Header />)

      return {
        enzymeWrapper
      }
    }

    it('should render self', () => {
      const { enzymeWrapper } = setup()

      expect(enzymeWrapper.find('h1').hasClass('title')).toBe(true)

      expect(enzymeWrapper.find('h1').text()).toBe('STARWARS')

      expect(enzymeWrapper.find('p').hasClass('subtitle')).toBe(true)

      expect(enzymeWrapper.find('p').text()).toBe('Search for your favorite star wars vehicles and starships and see their length in meters!')
    })

  })

  describe('Form', () => {
  let enzymeWrapper

   beforeEach(() => {
     enzymeWrapper = shallow(<ConnectedForm />)
   })

  
    it('should render self', () => {
      expect(toJson(enzymeWrapper)).toMatchSnapshot()
    })
  })

  describe('List', () => {
    let enzymeWrapper
  
     beforeEach(() => {
       enzymeWrapper = shallow(<ConnectedList />)
     })
  
    
      it('should render self', () => {
        expect(toJson(enzymeWrapper)).toMatchSnapshot()
      })
    })




})











