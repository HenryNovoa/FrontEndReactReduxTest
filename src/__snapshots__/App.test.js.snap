// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React components Form should render self 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <div
      className="control"
    >
      <input
        className="input"
        onChange={[Function]}
        placeholder="Vehicle or Spaceship"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="field"
  >
    <p
      className="control"
    >
      <span
        className="select"
      >
        <select
          onChange={[Function]}
        >
          <option
            value=""
          >
            Select
          </option>
        </select>
      </span>
    </p>
  </div>
  <div
    className="buttons"
  >
    <button
      className="button is-primary"
      disabled={true}
      type="submit"
    >
      Search
    </button>
    <button
      className="button is-link"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
</form>
`;

exports[`React components List should render self 1`] = `
<div
  className="section"
>
  <h1
    className="title"
  >
    Search Results
  </h1>
  <ul
    className="content"
  >
    <li
      className="box"
    >
      <div
        className="columns"
      >
        <div
          className="column"
        >
          No Search Results
        </div>
      </div>
    </li>
  </ul>
</div>
`;
